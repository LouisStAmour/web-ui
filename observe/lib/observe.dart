// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

/**
 * *Warning*: this library is experimental, and APIs are subject to change.
 *
 * This library is used to observe changes to [Observable] types. It also
 * has helpers to implement [Observable] objects.
 *
 * For example:
 *
 *     class Monster extends ObservableMixin {
 *       int _health = 100;
 *       static const _HEALTH = 'health';
 *       get health => _health;
 *       set health(value) {
 *         _health = notifyChange(_HEALTH, _health, value);
 *       }
 *       void damage(int amount) {
 *         print('$this takes $amount damage!');
 *         health -= amount;
 *       }
 *       toString() => 'Monster';
 *
 *       // These methods are temporary until dart2js supports mirrors.
 *       getValue(key) {
 *         if (key == _HEALTH) return health;
 *         return null;
 *       }
 *       setValue(key, val) {
 *         if (key == _HEALTH) health = val;
 *       }
 *     }
 *
 *     main() {
 *       var obj = new Monster();
 *       obj.changes.listen((records) {
 *         print('Changes to $obj were: $records');
 *       });
 *       // Asynchronously schedules delivery of these changes
 *       obj.damage(10);
 *       obj.damage(20);
 *       print('done!');
 *     }
 */
library observe;

import 'dart:async';
import 'dart:collection';
import 'dart:math' as math;

// TODO(jmesserly): use libraries instead of parts.
part 'src/list_diff.dart';
part 'src/observe_path.dart';
part 'src/observable.dart';
part 'src/observable_box.dart';
part 'src/observable_list.dart';
part 'src/observable_map.dart';
