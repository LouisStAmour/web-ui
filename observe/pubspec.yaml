name: observe
author: "Web UI Authors <web-ui-dev@dartlang.org>"
homepage: https://github.com/dart-lang/web-ui
description: observable objects for data binding
version: 0.0.1
dependencies:
  unittest: ">=0.4.6"
environment:
  sdk: ">=0.4.6"
