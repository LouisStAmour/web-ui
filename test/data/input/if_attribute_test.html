<!DOCTYPE html>
<!--
Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
for details. All rights reserved. Use of this source code is governed by a
BSD-style license that can be found in the LICENSE file.
-->
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <script src="packages/web_ui/testing/testing.js"></script>
</head>
<body>
  <!-- Tests the iterate attribute on a node. -->
  <div id='test' template if="x == 3">{{x}}</div>
  <script type="application/dart">
  import 'dart:async';
  import 'dart:html';
  import 'package:unittest/unittest.dart';
  import 'package:web_ui/observe.dart';
  @observable var x = 2;

  main() {
    Timer.run(() {
      expect(document.query('#test'), isNull);
      var placeholder = document.body.query('div#__e-2');
      expect(placeholder, isNotNull);
      expect(placeholder.style.display, equals('none'));
      x = 3;
      Timer.run(() {
        var div = document.query('#test');
        expect(div, isNotNull);
        expect(div.innerHtml, equals('3'));
        expect(div.previousNode, same(placeholder));
        window.postMessage('done', '*');
      });
    });
  }
  </script>
</body>
</html>
