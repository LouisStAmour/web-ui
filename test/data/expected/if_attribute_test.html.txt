Content-Type: text/plain
<html lang="en"><head><style>template { display: none; }</style>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <script src="../input/packages/web_ui/testing/testing.js"></script>
</head>
<body>
  <!-- Tests the iterate attribute on a node. -->
  <div style="display:none"></div><div id="test" template="" if="x == 3">3</div>
  


<script type="text/javascript" src="packages/web_ui/shadowdom.debug.js"></script>
<script type="application/dart" src="if_attribute_test.html_bootstrap.dart"></script></body></html>
